### PR TITLE
Get room nickname automatically

### PR DESCRIPTION
**Pull request by powdahound**
This makes it so Wobot fetches the room nickname it should use automatically. Users often have a hard time setting this properly or don't realize why they can't join rooms once they rename their bot. This makes things easy.

Changes of note:
- The 'connect' event is now emitted after the name has been fetched, not as soon as the client connects. Should there still be an event for when the initial connection is made?
- If there's an error fetching the profile info (HipChat server troubles) the bot just disconnects. I think this is the right behavior but I'm not sure the way I'm emitting an error and disconnecting is accurate. Can you confirm?

### CODE REVIEW

**Comment by powdahound** [comment]
Wow - the commit history gets a little crazy when merging between topic branches. :o Diff looks clean though!

**Comment by cjoudrey** [comment]
For the disconnect, I binded the private function `onError` to the XMPP `error` event.

You will notice that whenever this [function is called](https://github.com/powdahound/wobot/blob/8c9f7ca7622c84ca2dd0f36661873b4713bb0742/lib/bot.js#L14), the bot is disconnected.

The error is then emitted to user-land where plugins can catch it.

That being said, I think this might be better:

``` javascript
if (err) {
  onError('Unable to get profile info: ' + err);
  return;
}
```

That way there is a centralized area where all critical errors hit.

What do you think?

**Comment by cjoudrey** [comment]
As for the `emit('connect')`, that makes perfect sense.

I don't think we need another event for the time between bot connects and obtains profile. I don't really see a use for it.

**Comment by powdahound** [comment]
Hmm, calling `onError` in that way throws the following error: `Caught exception: TypeError: Object #<Object> has no method 'emit'` (the emit in the else case of onError). Know why that is? I'm just about to head out so can't dive in.

By the way, the things that `onError` currently handles are technically considered "stream errors" and disconnecting after they occur is the right thing to do. Receiving an error response (could be an iq, presence, or message stanza) from the server doesn't usually mean the connection should end. This nickname-fetching case seems to be an exception, so I'm not sure if it should use the stream error (onError) flows. In fact, maybe onError should actually be onStreamError.

**Comment by cjoudrey** [comment]
Ahhh.. Yeah... It's because `this` isn't binded to the instance of the bot.

That being said, you are entirely right. The private function `onError` should be called `onStreamError`. Once it's renamed, it becomes quite clear that we shouldn't be calling it for other errors such as failing to fetch profile.

That being said, the user-land API should still be called `onError`. All the user needs to know is, if `onError` is triggered, his bot will disconnect because something very bad happen. He can then handle the error however he wants (reconnect bot, alert someone, etc...)

For now what you did is just fine, but I've created the following issues:
- #19 - The bot disconnecting on error should be added via an event instead of hardcoded in the private API
- #20 - Private API - onError should be called onStreamError and only used by XMPP connection

Once #20 is done, your code will become:

``` javascript
self.emit('error', 'Unable to get profile info: ' + err);
return;
```

I'll get this done tonight and push `0.5.0` tonight or tomorrow depending on how many new features we have.

**Comment by powdahound** [comment]
Cool - I'll wait on those changes.

**Comment by cjoudrey** [comment]
I merged #19 and #20 as well as your two other pull requests.

You might get merge conflicts when you pull into your branch. I've been getting quite a few because of the new things in the `Bot` constructor.

**Comment by powdahound** [comment]
This is now ready for merge.

File: lib/bot.js [hunk 1]
```diff
@@ -57,19 +57,29 @@ module.exports.Bot = (function() {
...
       self.jabber.send(' ');
       self.emit('ping');
     }, 30000);
 
-    self.emit('connect');
+    // load our profile to get name
+    this.getProfile(function(err, data) {
+      if (err) {
+        self.emit('error', 'Unable to get profile info: ' + err);
+        self.disconnect();
```
Comment (cjoudrey) [inline lib/bot.js]: You can now remove this line as the `error` event will cause the bot to disconnect:

https://github.com/cjoudrey/wobot/blob/master/lib/bot.js#L165

**Comment by cjoudrey** [comment]
Awesome! I'll merge it when I get home. Thanks again!

**Pull request closed without merging**